### ISSUE TITLE
canu crashes at error correction. Reads too large?

### ISSUE THREAD

**Issue by conchoecia** [marbl/canu#475]
Hi there,

My current assembly is crashing during the falcon-sense stage. I got the 'don't worry, canu crashed  but it's probably not a big deal` error for the second time. Already tried to run the same command but no success.

I think it may due to some of the nanopore reads being too long ( 74MBp in 36 reads longer than 1MBp). 

Here's the final call for `falcon-sense` in `correction/2-correction/correctReads.sh`

```
$bin/falcon_sense \
  --min_idt 0.5 \
  --min_len 1000\
  --max_read_len 4194302 \
  --min_ovl_len 500\
  --min_cov 4 \
  --n_core 3 \
```

Looking at one of the failed jobs's log files, `correctReads.000089.out`

```
Running job 89 based on command line options.
Killed
Read layout generation failed.
mv: cannot stat './correction_outputs/0089.fasta': No such file or directory
```

I gave `canu` 90 threads on our assembly server, and the memory usage was peaking at 100% of 1TB at this point. It looked like every time the RAM peaked another job printed to STDOUT, making me think that these were some of the failed jobs. Around 30-40 of 125 total correction jobs failed.

I wasn't sure if the `--max_read_len 4194302` arg for `falcon-sense` was the default option, or if it was the length of my longest read in the fastq file. I checked and the longest 5 reads are:

```
3,643,134bp
3,904,152bp
4,373,757bp
5,953,647bp
10,235,264bp
```
Nope - not sure how the number for `--max_read_len 4194302` was determined. This doesn't match any of my longest read lengths.

How does falcon-sense determine the max read length and how can I avoid OOM crashes? For now I removed all reads less than 500bkp and am going to try again.

**Comment by skoren** [comment]
Wow, a 10mb read. Any idea if this read is real/aligned or noise? Currently Canu is limited to 2mb sequences so it will truncate these longest reads to the first 2mb. If there are routinely reads >2mb we can update that limit.

As for why the crash happened, it looks like the jobs were killed, most likely because too many concurrent processes were running, exhausting memory. If you are not using 1.5, definitely update. You can try limiting the concurrency explicitly (by setting ```corConcurrency=10```) and see if it completes. You should be able to re-run your original Canu command with the added parameter and it will rerun the failed jobs.

**Comment by conchoecia** [comment]
Thanks @skoren, I'll try out the `corConcurrency=10` option once this run is over to try to include the longer reads. I'll map the reads now and report back later this week if they are nonsensical or seem real.

**Comment by brianwalenz** [comment]
corConcurrency just limits the number of correction jobs running at the same time.  The hope is that if 40 jobs at once exhaust memory, maybe only 10 jobs at once will fit.  And if not 10, then maybe 5 or 2 or 1.

Increasing the read length limitation requires a (trivial) code change.  There's no real point in trying it until this memory issue gets figured out.

**Comment by eernst** [comment]
I'm seeing the same issue with corConcurrency=4 and nanopore "reads" > 1 Mpb. There are a number of such reads in my data causing multiple correction partitions to fail.

I've isolated an example of a correction read bundle (read# and data#s) with a 2 Mbp read that causes the error, which is:

`falcon_sense: falcon_sense/libfalcon/falcon.C:391: FConsensus::consensus_data* FConsensus::get_cns_from_align_tags(FConsensus::align_tags_t**, uint32, uint32, uint32, uint32): Assertion `t_len < max_len' failed.

Failed with 'Aborted'; backtrace (libbacktrace):
AS_UTL/AS_UTL_stackTrace.C::102 in _Z17AS_UTL_catchCrashiP9siginfo_tPv()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
falcon_sense/libfalcon/falcon.C::391 in _ZN10FConsensus23get_cns_from_align_tagsEPPNS_12align_tags_tEjjjj()
falcon_sense/libfalcon/falcon.C::699 in _ZN10FConsensus18generate_consensusESt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS6_EEjjdjj()
falcon_sense/falcon_sense.C::113 in main()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()`

It seems likely that the read is a chimera, possibly of several molecules passing through the pore in succession, but haven't confirmed this. 

If you'd like to take a look, the read bundle is here:
https://drive.google.com/file/d/0B81ZEZ28zd85SEFVZXVUbFVCS00/view?usp=sharing

**Comment by brianwalenz** [comment]
Awesome!  Not exactly the same though.  Can you open a new issue for this?  Thanks.

**Comment by skoren** [comment]
Reducing concurrent jobs along with the fix for issue #478 should address this bug. Open a new issue if you encounter other errors in this assembly.
